### PR TITLE
feat: Implement native version of ColumnarToRow

### LINKED ISSUE #708: Implement native version of ColumnarToRow
### What is the problem the feature request solves?

Spark's `ColumnarToRowExec` can be very slow in some cases. This plan shows that C2R took 8 minutes even though the underlying scan only took 20 seconds. This example is selecting decimal columns, so this is related to https://github.com/apache/datafusion-comet/issues/670 and https://github.com/apache/datafusion-comet/issues/679, but implementing a native version of `ColumnarToRowExec` may help with this.

image (3)

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by parthchandra**
## Which issue does this PR close?

Closes #708 

## Rationale for this change

Columnar to row operation is really slow and this attempts to speed it up.

## What changes are included in this PR?

Implements a native columnar to row operation for integer and floating point types. Variable length types can be added later.
The implementation allocates an off-heap buffer large enough to hold an entire batch of `UnsafeRow`s, passes it in to the native layer which converts the Arrow vectors into `UnsafeRow`s and puts them into the allocated buffer. The JVM code then simply points its `UnsafeRow`s to the underlying buffer.  
The native conversion copies the data from the arrow vector into the UnsafeRows. This is currently the performance bottleneck.

## How are these changes tested?
Additional unit test.

### CODE REVIEW

File: common/src/main/scala/org/apache/comet/CometConf.scala [hunk 1]
```diff
@@ -196,6 +196,12 @@ object CometConf extends ShimCometConf {
       .booleanConf
       .createWithDefault(false)
 
+  val COMET_EXEC_NATIVE_COLUMNAR_TO_ROW_ENABLED: ConfigEntry[Boolean] =
+    conf("spark.comet.exec.nativeColumnarToRow.enabled")
+      .doc("Experimental support for native columnar to row for fixed width types")
+      .booleanConf
+      .createWithDefault(true)
```
Comment (parthchandra) [inline common/src/main/scala/org/apache/comet/CometConf.scala:203]: Not the final version. Made it true default only to make sure all tests use this

File: common/src/main/scala/org/apache/spark/sql/comet/util/Utils.scala [hunk 2]
```diff
@@ -261,4 +262,32 @@ object Utils {
...
+        val dt = fromArrowField(v.getValueVector.getField)
+        assert(
+          UnsafeRow.isMutable(dt) || dt.isInstanceOf[BinaryType] || dt.isInstanceOf[StringType])
+        // For variable length types, assuming that the vector has not been read from, the
+        // readable bytes are the number of bytes of data in the vector.
+        val fixedBytes = num_rows * 8L // offset (4 bytes) and length (4 bytes)
+        val varBytes = dt match {
+          case datatype if UnsafeRow.isFixedLength(datatype) => 0L
+          case DecimalType.Fixed(_, _) => num_rows * 16L
+          case BinaryType | StringType =>
```
Comment (parthchandra) [inline common/src/main/scala/org/apache/spark/sql/comet/util/Utils.scala:282]: This case is currently unused since only the fixed length types are implemented. This will probably change completely.

**Comment by parthchandra** [comment]
Initial performance numbers for this implementation are not looking good. There are two areas where things are getting slower compared to Spark 
1 . No WholestageCodegen - The iteration over rows alone is adding extra cost to the implementation
2. We incur an additional cost of creating UnsafeRows in native which is more costly than the calls made by Spark to extract values out of Arrow vector.
Here's the initial benchmark run for just integer types - 
```
Running benchmark: ColumnarToRowExec
  Running case: Spark Columnar To Row - integer
  Stopped after 34 iterations, 2029 ms
  Running case: Comet Columnar To Row - integer
  Stopped after 24 iterations, 2081 ms

OpenJDK 64-Bit Server VM 11.0.19+7-LTS on Mac OS X 14.6
Apple M3 Max
ColumnarToRowExec:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark Columnar To Row - integer                      40             60          14        262.1           3.8       1.0X
Comet Columnar To Row - integer                      53             87          32        198.2           5.0       0.8X
```

**Comment by parthchandra** [comment]
@mbutrovich @andygrove any thoughts?

File: native/core/src/execution/shuffle/row.rs [hunk 3]
```diff
@@ -235,6 +250,143 @@ impl SparkUnsafeRow {
...
+        schema: Vec<ArrowDataType>,
+        arrays: Vec<ArrayRef>,
+        num_rows: usize,
+        num_cols: usize,
+        addr: usize,
+    ) {
+        let mut row_start_addr: usize = addr;
+        for i in 0..num_rows {
+            let mut row = SparkUnsafeRow::new(&schema);
+            let row_size = SparkUnsafeRow::get_row_bitset_width(schema.len()) + 8 * num_cols;
```
Comment (andygrove) [inline native/core/src/execution/shuffle/row.rs:264]: perhaps the compiler already optimizes this out, but this could happen outside of the loop?

File: native/core/src/execution/shuffle/row.rs [hunk 4]
```diff
@@ -271,6 +423,51 @@ impl SparkUnsafeRow {
...
+        unsafe {
+            let mask: i64 = 1i64 << (index & 0x3f);
+            let word_offset = (self.row_addr + (((index >> 6) as i64) << 3)) as *mut i64;
+            let word: i64 = *word_offset;
+            *word_offset = word | mask;
+        }
+    }
+
+    #[inline]
+    pub fn set_boolean(&mut self, index: usize, value: bool) {
```
Comment (andygrove) [inline native/core/src/execution/shuffle/row.rs:438]: These `set` methods could potentially be split out into an independent PR if we are unsure about merging this one

File: native/core/src/execution/shuffle/row.rs (hunk 3 above)
Comment (parthchandra) [reply native/core/src/execution/shuffle/row.rs:264]: I moved it out. However with variable length types, every row will be different in size so this will come back in.

File: native/core/src/execution/shuffle/row.rs (hunk 4 above)
Comment (parthchandra) [reply native/core/src/execution/shuffle/row.rs:438]: I can create a PR but I don't see the usefulness of doing so. It's looking like creating an `UnsafeRow` in native is a slower approach in general, so best avoided.

File: native/core/src/execution/shuffle/row.rs [hunk 5]
```diff
@@ -235,6 +250,143 @@ impl SparkUnsafeRow {
...
+                    row_start_addr as *const u8,
+                    row_size,
+                ));
+            }
+            row_start_addr += row_size;
+            for j in 0..num_cols {
+                let arr = arrays.get(j).unwrap();
+                let dt = &schema[j];
+                // assert_eq!(dt, arr.data_type());
+                match dt {
```
Comment (parthchandra) [inline native/core/src/execution/shuffle/row.rs:276]: This inner loop is the performance bottleneck. I moved the null check out of the match. It made a negligible difference. I also spent a bunch of time trying to eliminate the datatype check for every row and column, but the compiler wouldn't let me. Essentially, for every column I need an `ArrayAccessor` trait object that I can save in a vector so that I can call the `value(i)` (or better still `value_unchecked` method. I couldn't get the compiler to let me though.  Even if it did, this would make calls to the `value` method use dynamic dispatch and most likely be slower than the match on datatype.

Comment (Dandandan) [reply native/core/src/execution/shuffle/row.rs:276]: Some ideas you may want to try:
1. I think you could move the dynamica dispatch outside of the loop either with generics or macro (you want to generate the full inner loop on rows per data type). See 
https://arrow.apache.org/rust/arrow/array/trait.ArrayAccessor.html
for an example of how to use ArrayAccessor, you probably want to create a new function with a generic `T` with a bound on ArrayAccessor for the inner loop.
2. You could also avoid `is_null(i)` for array with `null_count` of 0 (if you do the check outside of the inner loop and write/generate two versions). This helps avoiding the check and often also will help generating better code.

**Comment by parthchandra** [comment]
@andygrove @Dandandan thanks for your comments. 
I've tried this now with a single datatype and no datatype checks and I see the same results. 
I'm going to close this PR until I can think of some better way to do this.

File: native/core/src/execution/shuffle/row.rs [hunk 6]
```diff
@@ -235,6 +250,143 @@ impl SparkUnsafeRow {
...
+    pub fn get_rows_from_arrays(
+        schema: Vec<ArrowDataType>,
+        arrays: Vec<ArrayRef>,
+        num_rows: usize,
+        num_cols: usize,
+        addr: usize,
+    ) {
+        let mut row_start_addr: usize = addr;
+        for i in 0..num_rows {
+            let mut row = SparkUnsafeRow::new(&schema);
```
Comment (andygrove) [inline native/core/src/execution/shuffle/row.rs:263]: This can be created once before the loop - we do not need to create once instance per row since this points to a memory address that gets updated each time we call `row.point_to_slice`

**Comment by andygrove** [comment]
@parthchandra I experimented with iterating over columns first and then rows. I also modified the code to only create `SparkUnsafeRow` once instead of once per row. I now see similar performance between Spark and native.

```
OpenJDK 64-Bit Server VM 11.0.24+8-post-Ubuntu-1ubuntu322.04 on Linux 6.8.0-47-generic
AMD Ryzen 9 7950X3D 16-Core Processor
ColumnarToRowExec:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark Columnar To Row - integer                      57             62           4        183.9           5.4       1.0X
Comet Columnar To Row - integer                      57             73          17        183.8           5.4       1.0X
```

Code:

```rust
                ArrowDataType::Int32 => {
                    let array = arr
                        .as_any()
                        .downcast_ref::<Int32Array>()
                        .expect("Error downcasting to Int32");

                    let mut row = SparkUnsafeRow::new(&schema);
                    let mut row_start_addr: usize = addr;

                    if array.null_count() == 0 {
                        for i in 0..num_rows {
                            let row_size =
                                SparkUnsafeRow::get_row_bitset_width(schema.len()) + 8 * num_cols;
                            unsafe {
                                row.point_to_slice(std::slice::from_raw_parts(
                                    row_start_addr as *const u8,
                                    row_size,
                                ));
                            }
                            row.set_int(j, array.value(i));
                            row_start_addr += row_size;
                        }

                    } else {
                        for i in 0..num_rows {
                            let row_size =
                                SparkUnsafeRow::get_row_bitset_width(schema.len()) + 8 * num_cols;
                            unsafe {
                                row.point_to_slice(std::slice::from_raw_parts(
                                    row_start_addr as *const u8,
                                    row_size,
                                ));
                            }
                            if array.is_null(i) {
                                row.set_null_at(j);
                            } else {
                                row.set_int(j, array.value(i));
                            }
                            row_start_addr += row_size;
                        }
                    }
                }
```

**Comment by parthchandra** [comment]
I considered going column then rows. This gets a little complicated with variable length as now we have to keep the offset for each row at which the variable length part can be written. 
This part of the code is no longer correct with variable length types - 
```
let row_size = SparkUnsafeRow::get_row_bitset_width(schema.len()) + 8 * num_cols;
```

**Comment by parthchandra** [comment]
@andygrove I tried with your change and I get - 
```
Running benchmark: ColumnarToRowExec
  Running case: Spark Columnar To Row - integer
  Stopped after 35 iterations, 2037 ms
  Running case: Comet Columnar To Row - integer
  Stopped after 15 iterations, 2022 ms

OpenJDK 64-Bit Server VM 11.0.19+7-LTS on Mac OS X 14.6
Apple M3 Max
ColumnarToRowExec:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Spark Columnar To Row - integer                      42             58          13        249.5           4.0       1.0X
Comet Columnar To Row - integer                      52            135          36        200.9           5.0       0.8X
```
So not really any better. I also tried with another change so that I precompute row lengths and rows start addresses to accomodate variable length fields and look up the row lengths as I iterate over the rows for each column. 
That slowed down to `0.7x` of Spark.
I don't think this is going to work, so I'm leaving this closed.
